lib/puppetserver/ca/sign_action.rb

```diff
@@ -0,0 +1,181 @@
+          opts.on('--help', 'Display this sign specific help output') do |help|
+            parsed['help'] = true
+          end
+          opts.on('--version', 'Output the version') do |v|
+            parsed['version'] = true
+          end
+          opts.on('--all', 'Operate on all certnames') do |a|
+            parsed['all'] = true
+          end
+        end
+      end
+
+      def parse(args)
+        results = { certname: []}
+        unparsed = []
+        parser = self.class.parser(results)
+        begin
+          parser.order!(args) do |unknown|
+            results[:certname] << unknown
+          end
+        rescue OptionParser::ParseError => e
+          unparsed += e.args
+          unparsed << args.shift unless args.first +~ /^-{1,2}/
+          retry
+        end
+        exit_code = nil
+
+        unless unparsed.empty?
+          @logger.err 'Error:'
+          @logger.err 'Unknown arguments or flags:'
+          unparsed.each do |arg|
+            @logger.err "    #{arg}"
+          end
+
+          @logger.err ''
+          @logger.err parser.help
+
+          exit_code = 1
+        end
+
+        if results[:certname].length == 0 && !results['all']
+          @logger.err 'A certname to sign is required'
+          @logger.err parser.help
+          exit_code = 1
+        elsif results[:certname].length > 1
+          @logger.err 'Only one certname is allowed at a time'
+          @logger.err 'Received certnames:'
+          results[:certname].each do |certname|
+            @logger.err "    #{certname}"
+          end
+          exit_code = 1
+        else
+          results[:certname] = results[:certname].first #this should be different
+        end
+
+        return results, exit_code
+      end
+
+      def parse_inputs(inputs)
+        parsed = {}
+        unparsed = []
+
+        parser = self.class.parser(parsed)
+
+        begin
+          parser.order!(inputs) do |nonopt|
+            unparsed << nonopt
+          end
+        rescue OptionParser::ParseError => e
+          unparsed += e.args
+          unparsed << inputs.shift unless inputs.first =~ /^-{1,2}/
+          retry
+        end
+
+        return parser, parsed, unparsed
+      end
+
+      def logger_output_put(result)
+        if result.is_a? Net::HTTPNoContent
+          @logger.inform "Signed Certificate for #{certname}"
+        elsif result.is_a? Net::HTTPNotFound
+          @logger.err 'Error'
+          @logger.err "    Could not find certificate for #{certname}"
+          return 1 #this makes sense because not all certs found, I think?
+        else
+          @logger.err 'Error - gross debug output:'
+          @logger.err "    #{result.inspect}"
+          return 1
+        end
+      end
+
+      def logger_output_get(result)
+        if result.is_a? Net::HTTPOK
+          return 0
+        elsif result.is_a? Net::HTTPNotFound
+          @logger.err 'Error'
+          @logger.err "   Could not find all certificates."
```

Comment:
this doesn't seem to be a very useful error message. How can I make it more clear?

Reply:
Does `certificate_statuses` really return 404 sometimes? I would assume it would return some mostly empty JSON. The code seems to indicated that if there are no files on disk, it attempts to render an empty map of JSON and returns that. But you should verify if you can: start up puppetserver, delete all the things in `ca/signed` and `ca/requests` (note that this state is also highly unlikely, since you should at least always have the puppet master's cert in `signed`), and query the endpoint. If it does just return an empty map, I'd do less error handling here and just make sure your later code can handle getting no results.

Reply:
That's a really good point; I'll dig into it further. At the moment, I think things will break if we get back a map with no unsigned certs, since ``requestedCertnames`` would be set to nil. I'll add some logic to handle that.